File: src/lib/exam/curriculum.ts
import type { CurriculumTaxonomy, Skill, SkillLevel } from "@/lib/types";
import taxonomyData from "../../../content/curriculum-taxonomy.json";

export const curriculum = taxonomyData as unknown as CurriculumTaxonomy;

/** Flat map of all skills keyed by skill_id */
export function getAllSkills(): Map<string, Skill> {
  const skills = new Map<string, Skill>();
  for (const domain of curriculum.domains) {
    for (const category of domain.skill_categories) {
      for (const skill of category.skills) {
        // Default to "foundations" for any skill missing the level field
        const enrichedSkill = skill.level ? skill : { ...skill, level: "foundations" as SkillLevel };
        skills.set(skill.skill_id, enrichedSkill);
      }
    }
  }
  return skills;
}

/** Get a single skill by ID, or undefined if not found */
export function getSkillById(skillId: string): Skill | undefined {
  return getAllSkills().get(skillId);
}

/** Get all skill IDs for a given domain */
export function getSkillIdsForDomain(domainId: string): string[] {
  const domain = curriculum.domains.find((d) => d.domain_id === domainId);
  if (!domain) return [];
  return domain.skill_categories.flatMap((cat) =>
    cat.skills.map((s) => s.skill_id)
  );
}

/** Get skill IDs for a domain, filtered by level */
export function getSkillIdsForDomainByLevel(
  domainId: string,
  level: SkillLevel
): string[] {
  const domain = curriculum.domains.find((d) => d.domain_id === domainId);
  if (!domain) return [];
  return domain.skill_categories.flatMap((cat) =>
    cat.skills
      .filter((s) => (s as Skill).level === level || (!('level' in s) && level === "foundations"))
      .map((s) => s.skill_id)
  );
}

/** Get all skills filtered by level */
export function getSkillsByLevel(level: SkillLevel): Map<string, Skill> {
  const allSkills = getAllSkills();
  const filtered = new Map<string, Skill>();
  for (const [id, skill] of Array.from(allSkills.entries())) {
    if (skill.level === level) {
      filtered.set(id, skill);
    }
  }
  return filtered;
}

/** Validate that all prerequisite references point to real skill IDs */
export function validatePrerequisites(): { valid: boolean; errors: string[] } {
  const allSkills = getAllSkills();
  const errors: string[] = [];

  for (const [skillId, skill] of Array.from(allSkills.entries())) {
    for (const prereqId of skill.prerequisite_skills) {
      if (!allSkills.has(prereqId)) {
        errors.push(
          `Skill "${skillId}" references unknown prerequisite "${prereqId}"`
        );
      }
    }
  }

  return { valid: errors.length === 0, errors };
}

File: python_app/app/lib/curriculum.py
"""Load and query the curriculum taxonomy."""
import json
import logging
from pathlib import Path
from functools import lru_cache
from typing import Optional

logger = logging.getLogger(__name__)

# Curriculum JSON structure mirrors TypeScript types
CURRICULUM_PATH = Path(__file__).parent.parent.parent / "content" / "curriculum-taxonomy.json"


@lru_cache(maxsize=1)
def _load_taxonomy() -> dict:
    with open(CURRICULUM_PATH, "r") as f:
        return json.load(f)


@lru_cache(maxsize=1)
def get_all_skills() -> dict[str, dict]:
    """Return a flat map of skill_id -> skill dict for all skills in all domains.

    The JSON structure is: domains → skill_categories → skills
    """
    taxonomy = _load_taxonomy()
    skills: dict[str, dict] = {}
    for domain in taxonomy.get("domains", []):
        for category in domain.get("skill_categories", []):
            for skill in category.get("skills", []):
                skill_id = skill["skill_id"]
                skills[skill_id] = skill
    return skills


def get_skill_by_id(skill_id: str) -> Optional[dict]:
    """Look up a single skill by ID. Returns None if not found."""
    return get_all_skills().get(skill_id)


def get_skill_ids_for_domain(domain_id: str) -> list[str]:
    """Return all skill IDs for a given domain."""
    taxonomy = _load_taxonomy()
    for domain in taxonomy.get("domains", []):
        if domain.get("domain_id") == domain_id:
            ids: list[str] = []
            for category in domain.get("skill_categories", []):
                for skill in category.get("skills", []):
                    ids.append(skill["skill_id"])
            return ids
    return []


def get_all_domain_ids() -> list[str]:
    """Return all domain IDs."""
    taxonomy = _load_taxonomy()
    return [d["domain_id"] for d in taxonomy.get("domains", [])]


def validate_prerequisites() -> list[str]:
    """Validate that all prerequisite_skills references exist. Returns list of errors."""
    skills = get_all_skills()
    errors: list[str] = []
    for skill_id, skill in skills.items():
        for prereq_id in skill.get("prerequisite_skills", []):
            if prereq_id not in skills:
                errors.append(f"Skill '{skill_id}' references unknown prereq '{prereq_id}'")
    return errors
